packages/ui-modal/src/Modal/index.tsx

```diff
@@ -182,6 +198,20 @@ class Modal extends Component<ModalProps, ModalState> {
     )
   }
 
+  captureDocumentClick: React.MouseEventHandler<any> = (event) => {
+    const { target } = event
+    const _dialogElement = findDOMNode(this._content) as Element
+    this._preventCloseOnDocumentClick =
+      event.button !== 0 || contains(_dialogElement, target as Node)
+  }
+
+  onClick: React.MouseEventHandler<any> = (event) => {
```

Comment:
This event handler checks if `shouldCloseOnDocumentClick` is `true` and the mouse pointer is pressed inside the popped-up Dialog.